linchpin/provision/roles/libvirt/tasks/provision_libvirt_node.yml

```diff
@@ -23,69 +27,279 @@
+    - ["{{ libvirt_resource_name }}"]
+    - "{{ res_count.stdout }}"
+  loop_control:
+    loop_var: definition
+  when: node_exists['failed'] is defined and uri_hostname == 'localhost' and res_def['cloud_config'] is defined
+
+- name: "Prepare cloud-config/meta-data remote"
+  template:
+    src: "templates/cloud-config/meta-data"
+    dest: "/tmp/vm-{{ definition[0] }}_{{ definition[1] }}/meta-data"
+  with_nested:
+    - ["{{ libvirt_resource_name }}"]
+    - "{{ res_count.stdout }}"
+  loop_control:
+    loop_var: definition
+  remote_user: "{{ res_def['remote_user'] | default('root') }}"
+  delegate_to: "{{ uri_hostname }}"
+  when: node_exists['failed'] is defined and uri_hostname != 'localhost' and res_def['cloud_config'] is defined
+
+- name: "localhost: Generate ci data cd image for cloud-init when cloud config is defined"
+  command: mkisofs -o /tmp/vm-{{ definition[0] }}_{{ definition[1] }}.iso -V cidata -r -J --quiet /tmp/vm-{{ definition[0] }}_{{ definition[1] }}/user-data /tmp/vm-{{ definition[0] }}_{{ definition[1] }}/meta-data
+  with_nested:
+    - ["{{ libvirt_resource_name }}"]
+    - "{{ res_count.stdout }}"
+  loop_control:
+    loop_var: definition
+  when: res_def['cloud_config'] is defined and node_exists['failed'] is defined and uri_hostname == 'localhost'
+
+- name: "remote_host: Generate ci data cd image for cloud-init when cloud config is defined"
+  command: mkisofs -o /tmp/vm-{{ definition[0] }}_{{ definition[1] }}.iso -V cidata -r -J --quiet /tmp/vm-{{ definition[0] }}_{{ definition[1] }}/user-data /tmp/vm-{{ definition[0] }}_{{ definition[1] }}/meta-data
+  with_nested:
+    - ["{{ libvirt_resource_name }}"]
+    - "{{ res_count.stdout }}"
+  loop_control:
+    loop_var: definition
+  remote_user: "{{ res_def['remote_user'] | default('root') }}"
+  delegate_to: "{{ uri_hostname }}"
+  when: res_def['cloud_config'] is defined and node_exists['failed'] is defined and uri_hostname != 'localhost'
+
+- name: "Install VM"
+  command: "virt-install --connect {{ definition[6] }} --import --name {{ definition[0] }}_{{ definition[7] }} --autostart --network bridge={{ definition[3] }},model=virtio --ram {{ definition[4] }} --disk path={{ definition[1] }}/{{ definition[0] }}_{{ definition[7] }}.{{ definition[2] }},format={{ definition[5] }},bus=virtio,cache=none --disk path=/tmp/vm-{{ definition[0] }}_{{ definition[7] }}.iso,device=cdrom --wait 10 --os-type=linux --nographics"
+  with_nested:
+    - ["{{ libvirt_resource_name }}"]
+    - ["{{ libvirt_image_path }}"]
+    - ["{{ img_src_ext }}"]
+    - ["{{ res_def['network_bridge'] | default('virbr0')  }}"]
+    - ["{{ res_def['memory'] | default(1024)  }}"]
+    - ["{{ res_def['format'] | default('qcow2')  }}"]
+    - ["{{ res_def['uri'] }}"]
+    - "{{ res_count.stdout }}"
+  loop_control:
+    loop_var: definition
+  become: yes
+  ignore_errors: yes
+  when: node_exists['failed'] is defined and res_def['cloud_config'] is defined
+
+- name: "Remove cloud-init cdrom "
```

Comment:
Seems that we're just shelling out to shell out here. Can you not use the virt module to accomplish this task? Additionally, does using virsh create yet another dependency?

Reply:
@herlo : virt module doesnot yet support change-media function . 
We could raise a PR in ansible to make it do so .

Reply:
Probably a good idea. Or maybe suggest a cloud-init module? Not sure here. Let me know what you decide.